pkg/eventhandler/drainer/drain.go

```diff
@@ -122,7 +122,7 @@ func (s *DrainEventHandler) evictPods(ctx context.Context, k8sclient kubernetes.
 		return microerror.Maskf(evictionInProgressError, "%d pods still pending eviction, waiting", left)
 	}
 
-	err := backoff.RetryNotify(o, backoff.NewConstant(15*time.Minute, 10*time.Second), backoff.NewNotifier(s.Logger, ctx))
+	err := backoff.RetryNotify(o, backoff.NewConstant(timeout, 10*time.Second), backoff.NewNotifier(s.Logger, ctx))
```

Comment:
Is this draining performed as part of the reconciliation loop (instead of background goroutine for example)?
If yes, how does this behave if there are multiple nodes receiving event around the same time?

- Does only the first node get drained?
- Is the draining sequential?
  - If sequential, could workload end up from drained node to another one which will be drained next and then repeating?
  - Is there anything in the reconciliation changing the state for nodes that has gone through draining?
    - When draining logic delays next reconciliation round, could the first nodes be in non-optimal state for prolonged period of time?

Reply:
This is deployed as a daemonset and each node drains itself. The event handlers are run in the main thread and the problem you are describing can happen if multiple events fire at the same time for the same node, but the result is the same (drain and delete node) so it doesn't change much

Reply:
Aaah! Right. I totally forgot about that. Everything is fine then :+1: